### ISSUE TITLE
allow keyword list to accept multiple tokens per single CLI option.

### ISSUE THREAD

**Issue by Ravencentric** [BrianPugh/cyclopts#31]
```py
from pathlib import Path
from typing import Annotated

from cyclopts import App, Parameter, validators

app = App()

@app.default() # type: ignore
def cli(
    *,
    path: Annotated[
        Path, 
        Parameter(
            help="directory or file",
            validator=validators.Path(exists=True))
        ] = Path.cwd(),
    config: Annotated[
        Path, 
        Parameter(
            name=["--config"], 
            help="config file", 
            validator=validators.Path(exists=True))
        ] = Path.cwd() / "config.yaml",
    args: Annotated[
        list[str], 
        Parameter(
            name="--args", 
            help="args")
        ] = [],
) -> None:
    print(f"{path = }, type = {type(path)}")
    print(f"{config = }, type = {type(config)}")
    print(f"{args = }, type = {type(args)}")

if __name__ == "__main__":
    app()
```

```py
> python .\test.py --args abc def geh
Traceback (most recent call last):
  File "C:\Users\raven\Documents\GitHub\test\test.py", line 36, in <module>
    app()
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\core.py", line 455, in __call__
    command, bound = self.parse_args(tokens, **kwargs)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\core.py", line 413, in parse_args
    command, bound, unused_tokens = self.parse_known_args(tokens, console=console)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\core.py", line 330, in parse_known_args
    bound, unused_tokens = create_bound_arguments(command, unused_tokens)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\bind.py", line 365, in create_bound_arguments
    p2c = parameter2cli(f)
          ^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\bind.py", line 72, in parameter2cli
    p2c.setdefault(iparam, [])
  File "C:\Users\raven\AppData\Local\Programs\Python\Python312\Lib\inspect.py", line 2842, in __hash__
    return hash((self._name, self._kind, self._annotation, self._default))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unhashable type: 'list'
```

**Comment by BrianPugh** [comment]
interesting! I'll check this in the morning; I'm stuck in a python3.8 mindset (at least until it reaches EOL at the end of 2024), so I always use `List[str]` (which should work).

**Comment by Ravencentric** [comment]
`List[str]` doesn't work either. For reference I'm on:
```yaml
> poetry env info

Virtualenv
Python:         3.11.7
Implementation: CPython
Path:           C:\Users\raven\Documents\GitHub\test\.venv
Executable:     C:\Users\raven\Documents\GitHub\test\.venv\Scripts\python.exe
Valid:          True

System
Platform:   win32
OS:         nt
Python:     3.11.7
Path:       C:\Users\raven\scoop\apps\python311\current
Executable: C:\Users\raven\scoop\apps\python311\current\python.exe
```

**Comment by BrianPugh** [comment]
oh sorry! lists works in a different way than you are expecting. It works like:

```
$ python .\test.py --args abc --args def --args geh
```

If the annotated parameter's arguments are provided positionally, they can all be together. Let me think about support this.

**Comment by BrianPugh** [comment]
I don't think we'd be able to support this without it become unnecessarily unintuitive. It would require the `--my-list` keyword to be the final argument, consuming all remaining tokens. If the user were to put `--my-list` in the middle of their CLI args, it would unexpectedly consume downstream `--other-options`.

Maybe we can revisit this for v2 if the desire for this feature is strong (it would require the `allow_leading_hyphen` App option described in #26 ).

**Comment by BrianPugh** [comment]
For additional context, (and not saying this is good/bad), Cyclopts [handles this the same way as Typer](https://typer.tiangolo.com/tutorial/multiple-values/multiple-options/).

**Comment by Ravencentric** [comment]
I think most people would expect `--args abc def geh` to work just as it does in `argparse` due to it's sheer popularity. It's also a more intuitive syntax over `--args abc --args def --args geh`

**Comment by Ravencentric** [comment]
```py
> python .\test.py --args abc --args def --args geh
Traceback (most recent call last):
  File "C:\Users\raven\Documents\GitHub\test\test.py", line 36, in <module>
    app()
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\core.py", line 482, in __call__
    command, bound = self.parse_args(tokens, **kwargs)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\core.py", line 440, in parse_args
    command, bound, unused_tokens = self.parse_known_args(tokens, console=console)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\core.py", line 355, in parse_known_args
    bound, unused_tokens = create_bound_arguments(
                           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\bind.py", line 373, in create_bound_arguments
    p2c = parameter2cli(f, default_parameter=default_parameter)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\raven\Documents\GitHub\test\.venv\Lib\site-packages\cyclopts\bind.py", line 74, in parameter2cli
    p2c.setdefault(iparam, [])
  File "C:\Users\raven\scoop\apps\python311\current\Lib\inspect.py", line 2799, in __hash__
    return hash((self.name, self.kind, self.annotation, self.default))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unhashable type: 'list'
```
Am I doing something wrong?

**Comment by BrianPugh** [comment]
I'm able to replicate; fixing this now.

**Comment by BrianPugh** [comment]
ok, so the causing issue is the bad practice of setting a mutable default to a variable `args = []`. The better practice (which works) is:

```
    args: Annotated[
        Optional[list[str][, 
        Parameter(
            name="--args", 
            help="args")
        ] = None,
        
    # then in the body
    if args is None:
        args = []
```

With that said, let me see what I can do.

**Comment by BrianPugh** [comment]
The mutable defaults issue has been fixed in v1.2.0.

I'm going to leave this issue open to continue discussion on allowing lists to consume multiple tokens. This functionality is already partially in Cyclopts, it's just not fleshed out enough to be useful to a user, yet. I'll start working on this feature after completing the `allow_leading_hyphen` feature described in #26. V2 will have a target release date of T1~T2 2024.

**Comment by BrianPugh** [comment]
I've updated the title to reflect the current issue.

**Comment by Ravencentric** [comment]
this code simply works in v2 without any issues!
```py
> python .\test.py --args abc def ghi
path = WindowsPath('C:/Users/raven/tmp'), type = <class 'pathlib.WindowsPath'>
config = WindowsPath('C:/Users/raven/tmp/config.yaml'), type = <class 'pathlib.WindowsPath'>
args = ['abc', 'def', 'ghi'], type = <class 'list'>
flag = False, type = <class 'bool'>
```

```py
> python .\test.py --args abc def ghi --flag
path = WindowsPath('C:/Users/raven/tmp'), type = <class 'pathlib.WindowsPath'>
config = WindowsPath('C:/Users/raven/tmp/config.yaml'), type = <class 'pathlib.WindowsPath'>
args = ['abc', 'def', 'ghi'], type = <class 'list'>
flag = True, type = <class 'bool'>
```

**Comment by BrianPugh** [comment]
I forgot I kind of addressed this 😅

I think (will need to double check tests) I need to slightly tweak the current logic, though. Currently I think the logic is:

If an Iterable keyword option (e.g. `--args`) is specified,
1. If `Parameter.allow_leading_hyphen=False` (default behavior), consume tokens until a hyphenated option is reached.
2. If `Parameter.allow_leading_hyphen=True`, consume all remaining tokens.

I think a better behavior would be:
1. If `Parameter.allow_leading_hyphen=False` (default behavior), consume tokens until a hyphenated option is reached (same as before).
2. If `Parameter.allow_leading_hyphen=False`, consume "1 element" (typically 1 token, but can be more for `Tuple` types; e.g. `Tuple[int, int]` would be 2 tokens) worth from the CLI.

In either case, if additional elements are specified as positional arguments, the positional-supplied values are **prepended** to the keyword-supplied values.

@Ravencentric what do you think about this proposed solution? Regardless, I need to update the Coercion Rules docs to reflect the current/proposed behavior.

**Comment by BrianPugh** [comment]
Actually, more I think about, the less I like my proposed idea. Maybe the current v2 behavior is best...

**Comment by Ravencentric** [comment]
If I'm understanding it correctly, what would happen if user does `tuple[int, ...]` (no fixed length) or `list[int]`. I find the current v2 behaviour intuitive, it's similar to argparse which is something most python devs are also familiar with

**Comment by BrianPugh** [comment]
Currently `Tuple[blah, ...]` is explicitly blocked, but I see no reason for blocking this anymore. I'll allow indetermine-length tuples (treated similarly as lists).

I agree, i'll leave the current v2 behavior, just add a few more tests to confirm functionality, and update docs.

**Comment by BrianPugh** [comment]
this should be resolved in `develop-v2` via #60

**Comment by BrianPugh** [comment]
closing with the release of v2. We can reopen this issue, or open a new one, should the need arise.
